pkg/query/flamegraph_arrow.go

```diff
@@ -1209,6 +1256,235 @@ func (fb *flamegraphBuilder) addRowValues(r *profile.RecordReader, row, sampleRo
+		appendDictionaryIndex(fb.mappingBuildID, trimmedMappingBuildIDIndices, te.row)
+		copyUint64BuilderValue(fb.builderLocationAddress, trimmedLocationAddress, te.row)
+		copyOptBooleanBuilderValue(fb.builderLocationFolded, trimmedLocationFolded, te.row)
+		copyOptInt64BuilderValue(fb.builderLocationLine, trimmedLocationLine, te.row)
+		copyOptInt64BuilderValue(fb.builderFunctionStartLine, trimmedFunctionStartLine, te.row)
+		appendDictionaryIndex(fb.functionName, trimmedFunctionNameIndices, te.row)
+		appendDictionaryIndex(fb.functionSystemName, trimmedFunctionSystemNameIndices, te.row)
+		appendDictionaryIndex(fb.functionFilename, trimmedFunctionFilenameIndices, te.row)
+		for i := range fb.labels {
+			appendDictionaryIndex(fb.labels[i], trimmedLabelsIndices[i], te.row)
+		}
+
+		// The following two will never be null.
+		cum := fb.builderCumulative.Value(te.row)
+		trimmedCumulative.Append(cum)
+		trimmedDiff.Append(fb.builderDiff.Value(te.row))
+
+		// This gets the newly inserted row's index.
+		// It is used further down as the children's parent value when added to the trimmingQueue.
+		row := trimmedCumulative.Len() - 1
+
+		// Add this new row as child to its parent if not the root row (index 0).
+		if row != 0 {
+			if len(trimmedChildren[te.parent]) == 0 {
+				trimmedChildren[te.parent] = []int{row}
+			} else {
+				trimmedChildren[te.parent] = append(trimmedChildren[te.parent], row)
+			}
+		}
+
+		cumThreshold := float32(cum) * threshold
+		for _, cr := range fb.children[te.row] {
+			if v := fb.builderCumulative.Value(cr); v > int64(cumThreshold) {
+				// this row is above the threshold, so we need to keep it
+				// add this row to the queue to check its children.
+				trimmingQueue.push(trimmingElement{row: cr, parent: row})
+			} else {
+				// this row is below the threshold, so we need to trim it
+				fb.trimmed += v
+			}
+		}
+	}
+
+	// Next we just keep the values in the dictionaries that we need after trimming.
+	var err error
+	trimmedMappingBuildIDIndicesArray := trimmedMappingBuildIDIndices.NewArray()
+	releasers = append(releasers, trimmedMappingBuildIDIndicesArray)
+
+	fb.mappingBuildID, err = compactDictionary(fb.pool, array.NewDictionaryArray(
+		fb.mappingBuildID.DataType(),
+		trimmedMappingBuildIDIndicesArray,
+		fb.mappingBuildID.Dictionary(),
+	))
+	if err != nil {
+		return err
+	}
+	trimmedMappingFileIndicesArray := trimmedMappingFileIndices.NewArray()
+	releasers = append(releasers, trimmedMappingFileIndicesArray)
+	fb.mappingFile, err = compactDictionary(fb.pool, array.NewDictionaryArray(
+		fb.mappingFile.DataType(),
+		trimmedMappingFileIndicesArray,
+		fb.mappingFile.Dictionary(),
+	))
+	if err != nil {
+		return err
+	}
+	trimmedFunctionNameIndicesArray := trimmedFunctionNameIndices.NewArray()
+	releasers = append(releasers, trimmedFunctionNameIndicesArray)
+	fb.functionName, err = compactDictionary(fb.pool, array.NewDictionaryArray(
```

Comment:
These types of assignments become a great vector for leaked allocations.  Because `fb.functionName` is already allocated and needs to be released and now it can't be. 

Furthermore with these direct assignments and a defered release, you need to have the deferred release check if the pointer is nil in the event the function panic'd or errored.